R/functions-util.R

```diff
@@ -255,4 +255,67 @@ utils.enableNeighbours <- function(x) {
     c(x, paste0(..., " [", date(), "]"))
 }
 
-setAs("logical", "factor", function(from, to) factor(from))
\ No newline at end of file
+setAs("logical", "factor", function(from, to) factor(from))
+
+#' The function aggregates `x` for `toBin` falling into bins defined
+#' by `breaks` using the `fun` function.
+#'
+#' @details
+#'
+#' This is a combination of the code from the former bin_Spectrum.
+#'
+#' @param x `numeric` with the values that should be binned.
+#'
+#' @param toBin `numeric`, same length than `x`, with values to be used for the
+#'     binning.
+#'
+#' @param binSize `numeric(1)` with the size of the bins.
+#'
+#' @param breaks `numeric` defining the breaks/bins.
+#'
+#' @param fun `function` to be used to aggregate values of `x` falling into the
+#'     bins defined by `breaks`.
+#'
+#' @return `list` with elements `x` and `mids` being the aggregated values
+#'     of `x` for values in `toBin` falling within each bin and the bin mid
+#'     points.
+#'
+#' @author Johannes Rainer, Sebastian Gibb
+#'
+#' @noRd
+.bin_values <- function(x, toBin, binSize = 1, breaks = seq(floor(min(toBin)),
+                                                            ceiling(max(toBin)),
+                                                            by = binSize),
+                        fun = max) {
+    if (length(x) != length(toBin))
+        stop("lengths of 'x' and 'toBin' have to match.")
+    fun <- match.fun(fun)
+    breaks <- .fix_breaks(breaks, range(toBin))
+    nbrks <- length(breaks)
+    idx <- findInterval(toBin, breaks)
+    ## Ensure that indices are within breaks.
+    idx[which(idx < 1L)] <- 1L
```

Comment:
```suggestion
    idx[idx < 1L] <- 1L
```
There was a time when subsetting was slower with logical than integer in R. These times are gone:
```r
library("microbenchmark")
set.seed(123)
breaks <- sort(sample(1e4, 1e3))
n <- length(breaks)
toBin <- 1:1e4
idx <- findInterval(toBin, breaks)

f1 <- function(x, n) {
    x[which(idx < 1L)] <- 1L
    x[which(idx >= n)] <- n - 1L
}
f2 <- function(x, n) {
    x[idx < 1L] <- 1L
    x[idx >= n] <- n - 1L
}
microbenchmark(f1(idx, n), f2(idx, n))
# Unit: microseconds
#        expr    min      lq     mean  median      uq     max neval
#  f1(idx, n) 87.386 90.5535 93.44552 92.0650 93.6125 164.635   100
#  f2(idx, n) 60.811 63.1915 65.29903 63.8785 65.2260 118.910   100
```

Reply:
I think the reason was more to handle `NA`s returned by `findInterval`.

Reply:
No. `findInterval` just returns `NA` if it gets `NA` as input and *mz* should never be `NA`. To be sure we could add an `if (anyNA(idx)) stop(...)` before the subsetting (I guess this will reduce the speed advantage). We could test for `NA` in `breaks` in `Spectra::bin` instead of `.peaks_bin` (than the breaks need to be tested just once instead of `n`-times).